geoutils/raster/delayed.py

```diff
@@ -0,0 +1,777 @@
+        """Get block projected footprints as a single geodataframe."""
+
+        geogrids = self.get_blocks_as_geogrids()
+        footprints = [gg.footprint_projected(crs=crs) if crs is not None else gg.footprint for gg in geogrids]
+
+        return pd.concat(footprints)
+
+
+def _chunks2d_from_chunksizes_shape(
+    chunksizes: tuple[int, int], shape: tuple[int, int]
+) -> tuple[tuple[int, ...], tuple[int, ...]]:
+    """Get tuples of chunk sizes for X/Y dimensions based on chunksizes and array shape."""
+
+    # Chunksize is fixed, except for the last chunk depending on the shape
+    chunks_y = tuple(
+        min(
+            chunksizes[0],
+            shape[0] - i * chunksizes[0],
+        )
+        for i in range(int(np.ceil(shape[0] / chunksizes[0])))
+    )
+    chunks_x = tuple(
+        min(
+            chunksizes[1],
+            shape[1] - i * chunksizes[1],
+        )
+        for i in range(int(np.ceil(shape[1] / chunksizes[1])))
+    )
+
+    return chunks_y, chunks_x
+
+
+def _combined_blocks_shape_transform(
+    sub_block_ids: list[dict[str, int]], src_geogrid: GeoGrid
+) -> tuple[dict[str, Any], list[dict[str, int]]]:
+    """Derive combined shape and transform from a subset of several blocks (for source input during reprojection)."""
+
+    # Get combined shape by taking min of X/Y starting indices, max of X/Y ending indices
+    all_xs, all_ys, all_xe, all_ye = ([b[s] for b in sub_block_ids] for s in ["xs", "ys", "xe", "ye"])
+    minmaxs = {"min_xs": np.min(all_xs), "max_xe": np.max(all_xe), "min_ys": np.min(all_ys), "max_ye": np.max(all_ye)}
+    combined_shape = (minmaxs["max_ye"] - minmaxs["min_ys"], minmaxs["max_xe"] - minmaxs["min_xs"])
+
+    # Shift source transform with start indexes to get the one for combined block location
+    combined_transform = src_geogrid.shift(xoff=minmaxs["min_xs"], yoff=-minmaxs["min_ys"]).transform
+
+    # Compute relative block indexes that will be needed to reconstruct a square array in the delayed function,
+    # by subtracting the minimum starting indices in X/Y
+    relative_block_indexes = [
+        {"r" + s1 + s2: b[s1 + s2] - minmaxs["min_" + s1 + "s"] for s1 in ["x", "y"] for s2 in ["s", "e"]}
+        for b in sub_block_ids
+    ]
+
+    combined_meta = {"src_shape": combined_shape, "src_transform": tuple(combined_transform)}
+
+    return combined_meta, relative_block_indexes
+
+
+@dask.delayed  # type: ignore
+def _delayed_reproject_per_block(
+    *src_arrs: tuple[NDArrayNum], block_ids: list[dict[str, int]], combined_meta: dict[str, Any], **kwargs: Any
+) -> NDArrayNum:
+    """
+    Delayed reprojection per destination block (also rebuilds a square array combined from intersecting source blocks).
+    """
+
+    # If no source chunk intersects, we return a chunk of destination nodata values
+    if len(src_arrs) == 0:
+        dst_arr = np.zeros(combined_meta["dst_shape"], dtype=np.dtype("float64"))
```

Comment:
why is dtype set to float64 here?

Reply:
Good point, I forgot this bit! Only needs float32 to return NaN, will be cast automatically to float64 if that's the source array type (which is not known here because there are no source arrays intersecting.